File: sample-objc/SendBird-iOS/GroupChannel/ViewCell/MemberListTableViewCell.m
//
//  MemberListTableViewCell.m
//  SendBird-iOS
//
//  Created by Jed Kyung on 10/4/16.
//  Copyright © 2016 SendBird. All rights reserved.
//

#import <AFNetworking/UIImageView+AFNetworking.h>

#import "MemberListTableViewCell.h"
#import "Constants.h"

@interface MemberListTableViewCell()

@property (weak, nonatomic) IBOutlet UIImageView *profileImageView;
@property (weak, nonatomic) IBOutlet UILabel *nicknameLabel;
@property (weak, nonatomic) IBOutlet UILabel *onlineDateLabel;

@property (strong, nonatomic) SBDUser *user;

@end

@implementation MemberListTableViewCell

+ (UINib *)nib
{
    return [UINib nibWithNibName:NSStringFromClass([self class]) bundle:[NSBundle bundleForClass:[self class]]];
}

+ (NSString *)cellReuseIdentifier
{
    return NSStringFromClass([self class]);
}

- (void)setModel:(SBDUser *)aUser {
    self.user = aUser;
    
    [self.profileImageView setImageWithURL:[NSURL URLWithString:self.user.profileUrl] placeholderImage:[UIImage imageNamed:@"img_profile"]];
    self.nicknameLabel.text = self.user.nickname;

    if (self.user.connectionStatus == SBDUserConnectionStatusOnline) {
        self.onlineDateLabel.text = @"Online";
        self.onlineDateLabel.textColor = [Constants memberOnlineTextColor];
    }
    else {
        // Last seen at
        NSDateFormatter *lastMessageDateFormatter = [[NSDateFormatter alloc] init];
        
        NSDate *lastMessageDate = nil;
        if ([NSString stringWithFormat:@"%lld", self.user.lastSeenAt].length == 10) {
            lastMessageDate = [NSDate dateWithTimeIntervalSince1970:(double)self.user.lastSeenAt];
        }
        else {
            lastMessageDate = [NSDate dateWithTimeIntervalSince1970:(double)self.user.lastSeenAt / 1000.0];
        }
        NSDate *currDate = [NSDate date];
        
        NSDateComponents *lastMessageDateComponents = [[NSCalendar currentCalendar] components:NSCalendarUnitDay | NSCalendarUnitMonth | NSCalendarUnitYear fromDate:lastMessageDate];
        NSDateComponents *currDateComponents = [[NSCalendar currentCalendar] components:NSCalendarUnitDay | NSCalendarUnitMonth | NSCalendarUnitYear fromDate:currDate];
        
        if (lastMessageDateComponents.year != currDateComponents.year || lastMessageDateComponents.month != currDateComponents.month || lastMessageDateComponents.day != currDateComponents.day) {
            [lastMessageDateFormatter setDateStyle:NSDateFormatterShortStyle];
            [lastMessageDateFormatter setTimeStyle:NSDateFormatterNoStyle];
            self.onlineDateLabel.text = [lastMessageDateFormatter stringFromDate:lastMessageDate];
        }
        else {
            [lastMessageDateFormatter setDateStyle:NSDateFormatterNoStyle];
            [lastMessageDateFormatter setTimeStyle:NSDateFormatterShortStyle];
            self.onlineDateLabel.text = [lastMessageDateFormatter stringFromDate:lastMessageDate];
        }
        
        self.onlineDateLabel.textColor = [Constants memberOfflineDateTextColor];
    }
}

@end

File: sample-swift/SendBird-iOS/GroupChannel/ViewCell/MemberListTableViewCell.swift
//
//  MemberListTableViewCell.swift
//  SendBird-iOS
//
//  Created by Jed Kyung on 10/10/16.
//  Copyright © 2016 SendBird. All rights reserved.
//

import UIKit
import SendBirdSDK
import AlamofireImage

class MemberListTableViewCell: UITableViewCell {
    @IBOutlet weak var profileImageView: UIImageView!
    @IBOutlet weak var nicknameLabel: UILabel!
    @IBOutlet weak var onlineDateLabel: UILabel!
    
    private var user: SBDUser!

    override func awakeFromNib() {
        super.awakeFromNib()
        // Initialization code
    }

    override func setSelected(_ selected: Bool, animated: Bool) {
        super.setSelected(selected, animated: animated)

        // Configure the view for the selected state
    }

    static func nib() -> UINib {
        return UINib(nibName: String(describing: self), bundle: Bundle(for: self))
    }
    
    static func cellReuseIdentifier() -> String {
        return String(describing: self)
    }
    
    func setModel(aUser: SBDUser) {
        self.user = aUser
        
        if self.user.profileUrl != nil {
            self.profileImageView.af_setImage(withURL: URL(string: self.user.profileUrl!)!, placeholderImage:UIImage(named: "img_profile"))
        }
        else {
            self.profileImageView.image = UIImage(named: "img_profile")
        }
        self.nicknameLabel.text = self.user.nickname
        
        if self.user.connectionStatus == SBDUserConnectionStatus.online {
            self.onlineDateLabel.text = "Online"
            self.onlineDateLabel.textColor = Constants.memberOnlineTextColor()
        }
        else {
            // Last seen at
            let lastMessageDateFormatter = DateFormatter()
            var lastMessageDate: Date?
            if String(format: "%lld", self.user.lastSeenAt).characters.count == 10 {
                lastMessageDate = Date(timeIntervalSince1970: Double(self.user.lastSeenAt))
            }
            else {
                lastMessageDate = Date(timeIntervalSince1970: Double(self.user.lastSeenAt) / 1000.0)
            }
            let currDate = Date()
            
            let lastMessageDateComponents = NSCalendar.current.dateComponents([.day, .month, .year], from: lastMessageDate! as Date)
            let currDateComponents = NSCalendar.current.dateComponents([.day, .month, .year], from: currDate as Date)
            
            if lastMessageDateComponents.year != currDateComponents.year || lastMessageDateComponents.month != currDateComponents.month || lastMessageDateComponents.day != currDateComponents.day {
                lastMessageDateFormatter.dateStyle = DateFormatter.Style.short
                lastMessageDateFormatter.timeStyle = DateFormatter.Style.none
                self.onlineDateLabel.text = lastMessageDateFormatter.string(from: lastMessageDate!)
            }
            else {
                lastMessageDateFormatter.dateStyle = DateFormatter.Style.none
                lastMessageDateFormatter.timeStyle = DateFormatter.Style.medium
                self.onlineDateLabel.text = lastMessageDateFormatter.string(from: lastMessageDate!)
            }
            
            self.onlineDateLabel.textColor = Constants.memberOfflineDateTextColor()
        }
    }
}
